File: 2021/day_01.cpp
#include <iostream>
#include <string>
#include <vector>

#include "../utils/input.hpp"

auto prepare_input(const std::vector<std::string>& input) {
  std::vector<int> result;

  for (auto line : input) {
    result.push_back(std::stoi(line));
  }

  return result;
}

auto part_one(const std::vector<std::string>& input) {
  auto depths = prepare_input(input);
  size_t result{0};

  for (size_t index{1}; index < depths.size(); ++index) {
    result += depths[index] > depths[index - 1];
  }

  return result;
}

auto part_two(const std::vector<std::string>& input) {
  auto depths = prepare_input(input);
  size_t result{0};

  for (size_t index{3}; index < depths.size(); ++index) {
    auto previous = depths[index - 3] + depths[index - 2] + depths[index - 1];
    auto current = depths[index - 2] + depths[index - 1] + depths[index];
    result += current > previous;
  }

  return result;
}

int main() {
  // std::filesystem::path input_path{"../2021/data/input_01_mock.txt"};
  std::filesystem::path input_path{"../2021/data/input_01.txt"};
  utils::Reader reader(input_path);
  auto input = reader.get_lines();

  std::cout << "The answer to part one is: " << part_one(input) << std::endl;
  std::cout << "The answer to part two is: " << part_two(input) << std::endl;

  return 0;
}

File: 2021/day_01.hs
-- import System.IO
-- import qualified Data.Text as T
-- import qualified Data.Text.Read
import Data.Char

buildPairs [] = []
buildPairs [x] = []
buildPairs (x:y:ys) = (x,y) : buildPairs (y:ys)

buildTriples [] = []
buildTriples [x] = []
buildTriples [x,y] = []
buildTriples (x:y:z:zs) = (x,y,z) : buildTriples (y:z:zs)

isIncreased :: [(Int,Int)] -> [Int]
isIncreased = map (\ x -> if uncurry (<) x then 1 else 0)

sumTriples (x,y,z) = x + y + z

partOne x = sum $ (isIncreased . buildPairs) x

partTwo x = map sumTriples (buildTriples x)

readInt :: String -> Int
readInt = read

main = do
    contents <- readFile "../2021/data/input_01.txt"
    let input = map readInt . words $ contents
    print $ partOne input

    print $ (partOne . partTwo) input

